File: gateway/src/main/java/ru/romanow/gateway/config/SecurityConfiguration.java
package ru.romanow.gateway.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
import org.springframework.security.core.userdetails.ReactiveUserDetailsService;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.server.SecurityWebFilterChain;

@Configuration
public class SecurityConfiguration {

    public static final String USER_NAME = "ronin";
    public static final String PASSWORD = "test";

    @Bean
    public SecurityWebFilterChain springSecurityFilterChain(ServerHttpSecurity http) {
        http.authorizeExchange(spec -> spec
                    .pathMatchers(HttpMethod.OPTIONS, "/**").permitAll()
                    .pathMatchers(HttpMethod.GET, "/manage/**").permitAll()
                    .pathMatchers("/dict/**").authenticated()
            )
            .httpBasic();

        return http.build();
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public ReactiveUserDetailsService userDetailsService() {
        final var passwordEncoder = passwordEncoder();
        final var userDetails = User.withUsername(USER_NAME)
                                    .passwordEncoder(passwordEncoder::encode)
                                    .password(PASSWORD)
                                    .authorities("ROLE_USER")
                                    .build();
        return new MapReactiveUserDetailsService(userDetails);
    }

}

File: gateway/src/main/java/ru/romanow/gateway/config/SecurityConfiguration.kt
package ru.romanow.gateway.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpMethod.GET
import org.springframework.http.HttpMethod.OPTIONS
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService
import org.springframework.security.core.userdetails.ReactiveUserDetailsService
import org.springframework.security.core.userdetails.User
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.server.SecurityWebFilterChain

@Configuration
class SecurityConfiguration {

    @Bean
    fun springSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain =
        http.authorizeExchange {
            it.pathMatchers(OPTIONS).permitAll()
            it.pathMatchers(GET, "/manage/**").permitAll()
            it.anyExchange().authenticated()
        }
            .httpBasic { }
            .csrf { it.disable() }
            .build()

    @Bean
    fun passwordEncoder() = BCryptPasswordEncoder()

    @Bean
    fun userDetailsService(): ReactiveUserDetailsService {
        val userDetails = User.withUsername(USER_NAME)
            .passwordEncoder { passwordEncoder().encode(it) }
            .password(PASSWORD)
            .authorities("ROLE_USER")
            .build()
        return MapReactiveUserDetailsService(userDetails)
    }

    companion object {
        const val USER_NAME = "program"
        const val PASSWORD = "test"
    }
}
